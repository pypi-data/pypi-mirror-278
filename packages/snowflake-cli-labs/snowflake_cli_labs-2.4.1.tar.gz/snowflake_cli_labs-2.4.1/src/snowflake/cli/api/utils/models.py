from __future__ import annotations

import os
from typing import Any, Dict


class EnvironWithDefinedDictFallback(Dict):
    def __getattr__(self, item):
        try:
            return self[item]
        except KeyError as e:
            raise AttributeError(e)

    def __getitem__(self, item):
        if item in os.environ:
            return os.environ[item]
        return super().__getitem__(item)

    def __contains__(self, item):
        return item in os.environ or super().__contains__(item)

    def update_from_dict(self, update_values: Dict[str, Any]):
        return super().update(update_values)
