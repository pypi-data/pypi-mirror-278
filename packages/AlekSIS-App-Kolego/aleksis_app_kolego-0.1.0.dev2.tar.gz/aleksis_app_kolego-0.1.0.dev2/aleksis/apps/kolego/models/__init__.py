from .absence import Absence, AbsenceReason  # noqa: F401
