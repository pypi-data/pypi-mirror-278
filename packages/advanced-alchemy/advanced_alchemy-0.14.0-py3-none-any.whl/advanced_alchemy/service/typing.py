"""Service object implementation for SQLAlchemy.

RepositoryService object is generic on the domain model type which
should be a SQLAlchemy model.
"""

from __future__ import annotations

from typing import (
    Any,
    Generic,
    Protocol,
    TypeVar,
    cast,
)

from typing_extensions import TypeAlias

from advanced_alchemy.filters import StatementFilter  # noqa: TCH001
from advanced_alchemy.repository.typing import ModelT  # noqa: TCH001

try:
    from msgspec import Struct, convert  # pyright: ignore[reportAssignmentType,reportUnusedImport]
except ImportError:  # pragma: nocover

    class Struct(Protocol):  # type: ignore[no-redef] # pragma: nocover
        """Placeholder Implementation"""

    def convert(*args: Any, **kwargs: Any) -> Any:  # type: ignore[no-redef] # noqa: ARG001 # pragma: nocover
        """Placeholder implementation"""
        return {}


try:
    from pydantic import BaseModel  # pyright: ignore[reportAssignmentType]
    from pydantic.type_adapter import TypeAdapter  # pyright: ignore[reportUnusedImport, reportAssignmentType]
except ImportError:  # pragma: nocover

    class BaseModel(Protocol):  # type: ignore[no-redef] # pragma: nocover
        """Placeholder Implementation"""

    T = TypeVar("T")  # pragma: nocover

    class TypeAdapter(Generic[T]):  # type: ignore[no-redef] # pragma: nocover
        """Placeholder Implementation"""

        def __init__(self, *args: Any, **kwargs: Any) -> None:  # pragma: nocover
            super().__init__()

        def validate_python(self, data: Any, *args: Any, **kwargs: Any) -> T:  # pragma: nocover
            """Stub"""
            return cast("T", data)


ModelDictT: TypeAlias = "dict[str, Any] | ModelT"
ModelDictListT: TypeAlias = "list[ModelT | dict[str, Any]] | list[dict[str, Any]]"
FilterTypeT = TypeVar("FilterTypeT", bound="StatementFilter")
ModelDTOT = TypeVar("ModelDTOT", bound="Struct | BaseModel")
PydanticModelDTOT = TypeVar("PydanticModelDTOT", bound="BaseModel")
StructModelDTOT = TypeVar("StructModelDTOT", bound="Struct")
