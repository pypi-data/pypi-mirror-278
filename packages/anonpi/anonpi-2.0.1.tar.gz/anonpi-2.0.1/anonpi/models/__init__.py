"""
This module contains the models for the main application.

Do not edit the classes manually or import them directly.
"""


from anonpi.models.main import Call ,RecordedCall , User
from anonpi.enums import Language
from anonpi.models.sub import TopupHistory , CallHistory , UserTopupRecords , UserCallRecords


# Path: models/main/call.py

dev = True

__slots__ = [
    'Call',
    'RecordedCall',
    'Language',
    'TopupRecord',
    'CallRecord',
    'UserTopupRecords',
    'UserCallRecords'
]
__all__ = [
    'Call',
    'RecordedCall',
    'Language',
    'TopupRecord',
    'CallRecord',
    'UserTopupRecords',
    'UserCallRecords'
]