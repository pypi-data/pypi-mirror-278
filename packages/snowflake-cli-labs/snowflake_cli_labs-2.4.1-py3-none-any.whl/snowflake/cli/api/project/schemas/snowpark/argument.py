from __future__ import annotations

from typing import Optional

from pydantic import Field
from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel


class Argument(UpdatableModel):
    name: str = Field(title="Name of the argument")
    arg_type: str = Field(
        title="Type of the argument", alias="type"
    )  # TODO: consider introducing literal/enum here
    default: Optional[str] = Field(title="Default value for an argument", default=None)
