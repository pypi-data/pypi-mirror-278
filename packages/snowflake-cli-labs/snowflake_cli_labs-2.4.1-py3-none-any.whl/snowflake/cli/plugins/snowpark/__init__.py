from snowflake.cli.plugins.snowpark.commands import app
from snowflake.cli.plugins.snowpark.package.commands import app as package_app

app.add_typer(package_app)
