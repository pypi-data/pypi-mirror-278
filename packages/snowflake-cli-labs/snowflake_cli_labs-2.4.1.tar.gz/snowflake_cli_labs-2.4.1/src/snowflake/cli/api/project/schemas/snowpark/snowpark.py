from __future__ import annotations

from typing import List, Optional

from pydantic import Field
from snowflake.cli.api.project.schemas.snowpark.callable import (
    FunctionSchema,
    ProcedureSchema,
)
from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel


class Snowpark(UpdatableModel):
    project_name: str = Field(title="Project identifier")
    stage_name: str = Field(title="Stage in which projectâ€™s artifacts will be stored")
    src: str = Field(title="Folder where your code should be located")
    functions: Optional[List[FunctionSchema]] = Field(
        title="List of functions defined in the project", default=[]
    )
    procedures: Optional[List[ProcedureSchema]] = Field(
        title="List of procedures defined in the project", default=[]
    )
