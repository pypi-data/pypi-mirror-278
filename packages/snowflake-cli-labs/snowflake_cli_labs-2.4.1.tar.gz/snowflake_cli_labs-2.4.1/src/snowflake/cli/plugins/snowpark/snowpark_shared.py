from __future__ import annotations

import click
import typer
from snowflake.cli.api.commands.flags import (
    deprecated_flag_callback,
    deprecated_flag_callback_enum,
)
from snowflake.cli.plugins.snowpark.models import YesNoAsk


def deprecated_allow_native_libraries_option(old_flag_name: str):
    return typer.Option(
        YesNoAsk.NO.value,
        old_flag_name,
        help="Allows native libraries, when using packages installed through PIP",
        hidden=True,
        callback=deprecated_flag_callback_enum(
            f"{old_flag_name} flag is deprecated. Use --allow-shared-libraries flag instead."
        ),
    )


def resolve_allow_shared_libraries_yes_no_ask(allow_shared_libraries: YesNoAsk) -> bool:
    if allow_shared_libraries == YesNoAsk.ASK:
        return click.confirm("Continue with package installation?", default=False)
    else:
        return allow_shared_libraries == YesNoAsk.YES


AllowSharedLibrariesOption: bool = typer.Option(
    False,
    "--allow-shared-libraries",
    help="Allows shared (.so) libraries, when using packages installed through PIP.",
)

DeprecatedCheckAnacondaForPyPiDependencies: bool = typer.Option(
    True,
    "--check-anaconda-for-pypi-deps/--no-check-anaconda-for-pypi-deps",
    "-a",
    help="""Checks if any of missing Anaconda packages dependencies can be imported directly from Anaconda. Valid values include: `true`, `false`, Default: `true`.""",
    hidden=True,
    callback=deprecated_flag_callback(
        "--check-anaconda-for-pypi-deps flag is deprecated. Use --ignore-anaconda flag instead."
    ),
)

IgnoreAnacondaOption: bool = typer.Option(
    False,
    "--ignore-anaconda",
    help="Does not lookup packages on Snowflake Anaconda channel.",
)

SkipVersionCheckOption: bool = typer.Option(
    False,
    "--skip-version-check",
    help="Skip comparing versions of dependencies between requirements and Anaconda.",
)

IndexUrlOption: str | None = typer.Option(
    None,
    "--index-url",
    help="Base URL of the Python Package Index to use for package lookup. This should point to "
    " a repository compliant with PEP 503 (the simple repository API) or a local directory laid"
    " out in the same format.",
    show_default=False,
)

ReturnsOption: str = typer.Option(
    ...,
    "--returns",
    "-r",
    help="Data type for the procedure to return.",
)

OverwriteOption: bool = typer.Option(
    False,
    "--overwrite",
    "-o",
    help="Replaces an existing procedure with this one.",
)
