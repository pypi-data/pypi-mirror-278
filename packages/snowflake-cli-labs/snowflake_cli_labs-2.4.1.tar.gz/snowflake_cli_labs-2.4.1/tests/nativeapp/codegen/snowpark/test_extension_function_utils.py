from textwrap import dedent

import pytest
import snowflake.cli.plugins.nativeapp.codegen.snowpark.extension_function_utils as ef_utils
from snowflake.cli.api.project.schemas.snowpark.argument import Argument
from snowflake.cli.plugins.nativeapp.codegen.snowpark.models import (
    NativeAppExtensionFunction,
)


@pytest.mark.parametrize(
    "function_type, expected",
    [
        ("procedure", "PROCEDURE"),
        ("function", "FUNCTION"),
        ("aggregate function", "AGGREGATE FUNCTION"),
        ("table function", "FUNCTION"),
    ],
)
def test_get_sql_object_type(
    function_type, expected, native_app_extension_function_raw_data
):
    native_app_extension_function_raw_data["type"] = function_type
    extension_fn = NativeAppExtensionFunction(**native_app_extension_function_raw_data)
    assert ef_utils.get_sql_object_type(extension_fn) == expected


def test_get_sql_argument_signature():
    arg = Argument(name="foo", type="int")
    assert ef_utils.get_sql_argument_signature(arg) == "foo int"

    arg = Argument(name="foo", type="int", default="42")
    assert ef_utils.get_sql_argument_signature(arg) == "foo int DEFAULT 42"


def test_get_qualified_object_name(native_app_extension_function):
    native_app_extension_function.name = "foo"
    native_app_extension_function.schema_name = None

    assert ef_utils.get_qualified_object_name(native_app_extension_function) == "foo"

    native_app_extension_function.name = "foo"
    native_app_extension_function.schema_name = "my_schema"

    assert (
        ef_utils.get_qualified_object_name(native_app_extension_function)
        == "my_schema.foo"
    )

    native_app_extension_function.name = "foo"
    native_app_extension_function.schema_name = "my schema"

    assert (
        ef_utils.get_qualified_object_name(native_app_extension_function)
        == '"my schema".foo'
    )

    native_app_extension_function.name = "foo"
    native_app_extension_function.schema_name = "my.full.schema"

    assert (
        ef_utils.get_qualified_object_name(native_app_extension_function)
        == "my.full.schema.foo"
    )

    native_app_extension_function.name = "foo"
    native_app_extension_function.schema_name = "my.full schema.with special chars"

    assert (
        ef_utils.get_qualified_object_name(native_app_extension_function)
        == 'my."full schema"."with special chars".foo'
    )


def test_get_function_type_signature_for_grant(native_app_extension_function):
    assert (
        ef_utils.get_function_type_signature_for_grant(native_app_extension_function)
        == "int"
    )

    native_app_extension_function.signature = []
    assert (
        ef_utils.get_function_type_signature_for_grant(native_app_extension_function)
        == ""
    )

    native_app_extension_function.signature = [
        Argument(name="foo", type="int", default="42"),
        Argument(name="bar", type="varchar"),
    ]
    assert (
        ef_utils.get_function_type_signature_for_grant(native_app_extension_function)
        == "int, varchar"
    )


def test_ensure_string_literal():
    assert ef_utils.ensure_string_literal("") == "''"
    assert ef_utils.ensure_string_literal("abc") == "'abc'"
    assert ef_utils.ensure_string_literal("'abc'") == "'abc'"
    assert ef_utils.ensure_string_literal("'abc def'") == "'abc def'"
    assert ef_utils.ensure_string_literal("'abc") == r"'\'abc'"
    assert ef_utils.ensure_string_literal("abc'") == r"'abc\''"


def test_ensure_all_string_literals():
    assert ef_utils.ensure_all_string_literals([]) == []
    assert ef_utils.ensure_all_string_literals(["", "foo", "'bar'"]) == [
        "''",
        "'foo'",
        "'bar'",
    ]


TEST_SNOWPARK_CODE = dedent(
    """
import snowflake.snowpark

from typing import Optional
from snowflake.snowpark.session import Session
from snowflake.snowpark.functions import col, sum, sproc


@custom
@sproc(native_app_params={'schema': 'math', 'application_roles': ['app_public', 'app_admin']})
def sproc_sum(session: Session, first: int, second: int) -> int:
    return first + second

@udf(native_app_params={
    'schema': 'math',
    'application_roles': ['app_public', 'app_admin']
})
@module.annotation
def udf_sum(first: int, second: int) -> int:
    return first + second
    
@custom
def helper():
    pass
"""
).strip()


def test_deannotate_module_source_removes_all_annotations(snapshot):
    sproc = NativeAppExtensionFunction(
        type="procedure",
        handler="math_fns.sproc_sum",
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )
    udf = NativeAppExtensionFunction(
        type="function",
        handler="math_fns.udf_sum",
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )
    assert (
        ef_utils.deannotate_module_source(TEST_SNOWPARK_CODE, [sproc, udf]) == snapshot
    )


def test_deannotate_module_source_preserves_specified_annotations(snapshot):
    sproc = NativeAppExtensionFunction(
        type="procedure",
        handler="math_fns.sproc_sum",
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )
    udf = NativeAppExtensionFunction(
        type="function",
        handler="math_fns.udf_sum",
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )
    assert (
        ef_utils.deannotate_module_source(
            TEST_SNOWPARK_CODE, [sproc, udf], annotations_to_preserve=["custom"]
        )
        == snapshot
    )
    assert (
        ef_utils.deannotate_module_source(
            TEST_SNOWPARK_CODE,
            [sproc, udf],
            annotations_to_preserve=["module.annotation"],
        )
        == snapshot
    )


def test_deannotate_module_source_is_identity_when_no_functions_present(snapshot):
    sproc = NativeAppExtensionFunction(
        type="procedure",
        handler="math_fns.sproc_sum",
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )
    udf = NativeAppExtensionFunction(
        type="function",
        handler="math_fns.udf_sum",
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )

    non_annotated_code = dedent(
        """
    if __name__ == "__main__":
        print("All your base are belong to us")
    """
    )

    assert (
        ef_utils.deannotate_module_source(
            non_annotated_code, [sproc, udf], annotations_to_preserve=["custom"]
        )
        == non_annotated_code
    )


def test_deannotate_module_source_is_identity_when_no_annotated_functions_present(
    snapshot,
):
    sproc = NativeAppExtensionFunction(
        type="procedure",
        handler="math_fns.sproc_sum",
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )
    udf = NativeAppExtensionFunction(
        type="function",
        handler="math_fns.udf_sum",
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )

    non_annotated_code = dedent(
        """
    def foo():
        pass
    """
    )

    assert (
        ef_utils.deannotate_module_source(
            non_annotated_code, [sproc, udf], annotations_to_preserve=["custom"]
        )
        == non_annotated_code
    )


def test_deannotate_module_source_is_identity_when_extension_function_does_not_match(
    snapshot,
):
    sproc = NativeAppExtensionFunction(
        type="procedure",
        handler="math_fns.my_sproc_sum",  # trigger mismatch
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )
    udf = NativeAppExtensionFunction(
        type="function",
        handler="math_fns.my_udf_sum",  # trigger mismatch
        returns="int",
        signature=[
            Argument(name="first", type="int"),
            Argument(name="second", type="int"),
        ],
    )

    assert (
        ef_utils.deannotate_module_source(TEST_SNOWPARK_CODE, [sproc, udf])
        == TEST_SNOWPARK_CODE
    )
