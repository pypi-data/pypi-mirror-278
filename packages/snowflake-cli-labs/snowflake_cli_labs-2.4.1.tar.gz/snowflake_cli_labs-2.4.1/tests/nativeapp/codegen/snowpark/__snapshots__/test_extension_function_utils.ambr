# serializer version: 1
# name: test_deannotate_module_source_preserves_specified_annotations
  '''
  import snowflake.snowpark
  
  from typing import Optional
  from snowflake.snowpark.session import Session
  from snowflake.snowpark.functions import col, sum, sproc
  
  
  @custom
  #: @sproc(native_app_params={'schema': 'math', 'application_roles': ['app_public', 'app_admin']})
  def sproc_sum(session: Session, first: int, second: int) -> int:
      return first + second
  
  #: @udf(native_app_params={
  #:     'schema': 'math',
  #:     'application_roles': ['app_public', 'app_admin']
  #: })
  #: @module.annotation
  def udf_sum(first: int, second: int) -> int:
      return first + second
  
  @custom
  def helper():
      pass
  '''
# ---
# name: test_deannotate_module_source_preserves_specified_annotations.1
  '''
  import snowflake.snowpark
  
  from typing import Optional
  from snowflake.snowpark.session import Session
  from snowflake.snowpark.functions import col, sum, sproc
  
  
  #: @custom
  #: @sproc(native_app_params={'schema': 'math', 'application_roles': ['app_public', 'app_admin']})
  def sproc_sum(session: Session, first: int, second: int) -> int:
      return first + second
  
  #: @udf(native_app_params={
  #:     'schema': 'math',
  #:     'application_roles': ['app_public', 'app_admin']
  #: })
  @module.annotation
  def udf_sum(first: int, second: int) -> int:
      return first + second
  
  @custom
  def helper():
      pass
  '''
# ---
# name: test_deannotate_module_source_removes_all_annotations
  '''
  import snowflake.snowpark
  
  from typing import Optional
  from snowflake.snowpark.session import Session
  from snowflake.snowpark.functions import col, sum, sproc
  
  
  #: @custom
  #: @sproc(native_app_params={'schema': 'math', 'application_roles': ['app_public', 'app_admin']})
  def sproc_sum(session: Session, first: int, second: int) -> int:
      return first + second
  
  #: @udf(native_app_params={
  #:     'schema': 'math',
  #:     'application_roles': ['app_public', 'app_admin']
  #: })
  #: @module.annotation
  def udf_sum(first: int, second: int) -> int:
      return first + second
  
  @custom
  def helper():
      pass
  '''
# ---
